### PR TITLE
✨ Adds `AvailableCondition` and `ReadyCondition` Conditions to MachineDeployment

### PR DESCRIPTION
**Pull request by Arvinderpal**
**What this PR does / why we need it**:
Following the model of KCP and Machine Conditions, this PR adds an `AvailableCondition` to MachineDeployment. The condition is true when Nodes of the underlying MachineSet(s) are available to take on workloads. It also adds the summary `ReadyCondition`. 

Tracking issue: https://github.com/kubernetes-sigs/cluster-api/issues/3486

### CODE REVIEW

File: api/v1alpha3/zz_generated.conversion.go [hunk 1]
```diff
@@ -826,9 +826,11 @@ func autoConvert_v1alpha4_MachineDeploymentStatus_To_v1alpha3_MachineDeploymentS
 	out.AvailableReplicas = in.AvailableReplicas
 	out.UnavailableReplicas = in.UnavailableReplicas
 	out.Phase = in.Phase
+	// WARNING: in.Conditions requires manual conversion: does not exist in peer-type
 	return nil
 }
 
+// FIXME(awander): This func is removed after running `make generate`. I assume b/c there is no Conditions fields in v1alpha3 for MachineDeploymentStatus. What is the correct way to handle this?
```
Comment (Arvinderpal) [inline api/v1alpha3/zz_generated.conversion.go:833]: Can someone help me here. What is the correct way to handle this?

File: controllers/machinedeployment_sync.go [hunk 2]
```diff
@@ -349,6 +350,14 @@ func (r *MachineDeploymentReconciler) scale(ctx context.Context, deployment *clu
 // syncDeploymentStatus checks if the status is up-to-date and sync it if necessary.
 func (r *MachineDeploymentReconciler) syncDeploymentStatus(allMSs []*clusterv1.MachineSet, newMS *clusterv1.MachineSet, d *clusterv1.MachineDeployment) error {
 	d.Status = calculateStatus(allMSs, newMS, d)
+
```
Comment (enxebre) [inline controllers/machinedeployment_sync.go:353]: may be add a comment here that `minReplicasNeeded` will be equal to `d.Spec.Replicas` when the strategy is not `RollingUpdateMachineDeploymentStrategyType`

File: api/v1alpha4/condition_consts.go [hunk 3]
```diff
@@ -201,3 +201,14 @@ const (
...
 )
+
+// Conditions and condition Reasons for  MachineDeployments
+
+const (
+	// AvailableCondition means the MachineDeployment is available, ie. at least the minimum available
+	// machines required are up and running for at least minReadySeconds.
+	AvailableCondition ConditionType = "Available"
+
+	// MinimumMachinesAvailableReason reflects the minimum availability of machines for a machinedeployment.
```
Comment (enxebre) [inline api/v1alpha4/condition_consts.go:212]: I'm having a hard time to read this, "reflects the minimum availability of machines for a machinedeployment." doesn't seem right to me, since this variable is just meant to be the reason for AvailableCondition=false. Also shouldn't it be `NotMinimumMachinesAvailable`?

File: controllers/machinedeployment_controller_test.go [hunk 4]
```diff
@@ -389,6 +389,21 @@ var _ = Describe("MachineDeployment Reconciler", func() {
 			return len(machineSets.Items)
 		}, timeout*5).Should(BeEquivalentTo(0))
 
+		By("Verifying MachineDeployment has correct Conditions")
+		Eventually(func() bool {
+			key := client.ObjectKey{Name: deployment.Name, Namespace: deployment.Namespace}
+			if err := testEnv.Get(ctx, key, deployment); err != nil {
+				return false
+			}
+			var availableCond bool
```
Comment (enxebre) [inline controllers/machinedeployment_controller_test.go:398]: why do we need `availableCond` at all? can't we just return true/false below?

File: controllers/machinedeployment_controller_test.go [hunk 5]
```diff
@@ -389,6 +389,21 @@ var _ = Describe("MachineDeployment Reconciler", func() {
 			return len(machineSets.Items)
 		}, timeout*5).Should(BeEquivalentTo(0))
 
+		By("Verifying MachineDeployment has correct Conditions")
```
Comment (enxebre) [inline controllers/machinedeployment_controller_test.go:392]: Can we also try to include a negative test?

File: api/v1alpha4/condition_consts.go [hunk 6]
```diff
@@ -201,3 +201,14 @@ const (
 	// from making any further remediations.
 	TooManyUnhealthyReason = "TooManyUnhealthy"
 )
+
+// Conditions and condition Reasons for  MachineDeployments
+
+const (
+	// AvailableCondition means the MachineDeployment is available, ie. at least the minimum available
+	// machines required are up and running for at least minReadySeconds.
+	AvailableCondition ConditionType = "Available"
```
Comment (enxebre) [inline api/v1alpha4/condition_consts.go:210]: If we want this to be machineDeployment specific should probably rename the variable to `MachineDeploymentAvailableCondition` consistently with `ControlPlaneReadyCondition` or `BootstrapReadyCondition`. Otherwise it should probably live in this block https://github.com/kubernetes-sigs/cluster-api/pull/4625/files#diff-9bdb2e3a1ad45026b8684d8d8dab63c66ac79269293891a64b6684451356764dL22

File: controllers/machinedeployment_controller.go [hunk 7]
```diff
@@ -148,6 +154,26 @@ func (r *MachineDeploymentReconciler) Reconcile(ctx context.Context, req ctrl.Re
 	return result, err
 }
 
+func patchMachineDeployment(ctx context.Context, patchHelper *patch.Helper, d *clusterv1.MachineDeployment, options ...patch.Option) error {
+	// Always update the readyCondition by summarizing the state of other conditions.
+	conditions.SetSummary(d,
+		conditions.WithConditions(
+			clusterv1.AvailableCondition,
```
Comment (enxebre) [inline controllers/machinedeployment_controller.go:161]: As it is now does this mean there's no actual semantic difference between available and ready? are we planning to follow up with more conditions here?

File: api/v1alpha4/condition_consts.go (hunk 6 above)
Comment (Arvinderpal) [reply api/v1alpha4/condition_consts.go:210]: I'm happy to change that but I'm following the convention used for KCP where the condition is simply called`AvailableCondition`. Similarly, the `ReadyCondition` is used for several objects.

File: controllers/machinedeployment_controller.go (hunk 7 above)
Comment (Arvinderpal) [reply controllers/machinedeployment_controller.go:161]: Yes, I am. Please see the [tracking issue](https://github.com/kubernetes-sigs/cluster-api/issues/3486).

File: controllers/machinedeployment_controller_test.go (hunk 4 above)
Comment (Arvinderpal) [reply controllers/machinedeployment_controller_test.go:398]: I saw this convention used in other the places and used it here as other conditions will be added and checked.

File: controllers/machinedeployment_controller_test.go (hunk 5 above)
Comment (Arvinderpal) [reply controllers/machinedeployment_controller_test.go:392]: Yes, I can do that. I will likely add it under machinedeployment_sync_test.go

File: controllers/machinedeployment_sync.go (hunk 2 above)
Comment (Arvinderpal) [reply controllers/machinedeployment_sync.go:353]: So the only other strategy I see implemented today is `OnDeleteMachineDeploymentStrategyType`. Would the comment become stale if we implement other strategies?

**Review by Arvinderpal — COMMENTED** [review]
@enxebre Thank you for the quick feedback. I will work on adding a unit test for the negative case.

File: controllers/machinedeployment_controller.go [hunk 8]
```diff
@@ -148,6 +154,26 @@ func (r *MachineDeploymentReconciler) Reconcile(ctx context.Context, req ctrl.Re
...
+		),
+	)
+
+	// Patch the object, ignoring conflicts on the conditions owned by this controller.
+	// Also, if requested, we are adding additional options like e.g. Patch ObservedGeneration when issuing the
+	// patch at the end of the reconcile loop.
```
Comment (fabriziopandini) [inline controllers/machinedeployment_controller.go:167]: What about removing this because this is not the case now

File: controllers/machinedeployment_sync.go [hunk 9]
```diff
@@ -349,6 +350,14 @@ func (r *MachineDeploymentReconciler) scale(ctx context.Context, deployment *clu
...
+	minReplicasNeeded := *(d.Spec.Replicas) - mdutil.MaxUnavailable(*d)
+	if d.Status.AvailableReplicas >= minReplicasNeeded {
+		// NOTE: The structure of calculateStatus() does not allow us to update the machinedeployment directly, we can only update the status obj it returns. Ideally, we should change calculateStatus() --> updateStatus() to be consistent with the rest of the code base, until then, we update conditions here.
```
Comment (fabriziopandini) [inline controllers/machinedeployment_sync.go:356]: Should we fix this now?

File: controllers/machinedeployment_sync.go [hunk 10]
```diff
@@ -349,6 +350,14 @@ func (r *MachineDeploymentReconciler) scale(ctx context.Context, deployment *clu
...
+		conditions.MarkTrue(d, clusterv1.AvailableCondition)
+	} else {
+		conditions.MarkFalse(d, clusterv1.AvailableCondition, clusterv1.MinimumMachinesAvailableReason, clusterv1.ConditionSeverityWarning, "MachineDeployment does not have minimum availability of %d replicas, current availability is %d", minReplicasNeeded, d.Status.AvailableReplicas)
```
Comment (fabriziopandini) [inline controllers/machinedeployment_sync.go:359]:
```suggestion
		conditions.MarkFalse(d, clusterv1.AvailableCondition, clusterv1.MinimumMachinesAvailableReason, clusterv1.ConditionSeverityWarning, "Minimum availability requires %d replicas, current %d available", minReplicasNeeded, d.Status.AvailableReplicas)
```
Let's avoid MachineDeployment given that those conditions are already inside the MD  object.

File: api/v1alpha3/zz_generated.conversion.go (hunk 1 above)
Comment (fabriziopandini) [reply api/v1alpha3/zz_generated.conversion.go:833]: If I got this right you should make sure to get roundtrip conversion by using marshal and unamarshal to in an annotation like we already have for `spec.Strategy` in https://github.com/kubernetes-sigs/cluster-api/blob/440a0035c0ddf463147652976878cc2015dbc235/api/v1alpha3/conversion.go#L119-L159

File: api/v1alpha4/condition_consts.go [hunk 11]
```diff
@@ -201,3 +201,14 @@ const (
 	// from making any further remediations.
 	TooManyUnhealthyReason = "TooManyUnhealthy"
 )
+
+// Conditions and condition Reasons for  MachineDeployments
+
+const (
+	// AvailableCondition means the MachineDeployment is available, ie. at least the minimum available
+	// machines required are up and running for at least minReadySeconds.
```
Comment (fabriziopandini) [inline api/v1alpha4/condition_consts.go:209]: What about adding the formula for minus machine available

File: api/v1alpha4/condition_consts.go (hunk 6 above)
Comment (fabriziopandini) [reply api/v1alpha4/condition_consts.go:210]: Given this condition has a specific semantic, let use a different constant (Ready condition has the same semantic across the entire board)

File: controllers/machinedeployment_controller_test.go (hunk 5 above)
Comment (Arvinderpal) [reply controllers/machinedeployment_controller_test.go:392]: Done

**Review by Arvinderpal — COMMENTED** [review]
@fabriziopandini I have addressed all the comments. PTAL

**Comment by Arvinderpal** [comment]
Can anyone help me with why `pull-cluster-api-test-main ` is failing with: 
```
FAIL	sigs.k8s.io/cluster-api/controllers [build failed]
```
Is it possibly related to this line in build logs? I did add a new test at line 393, but I don't quite understand why that would be the issue: 

```# sigs.k8s.io/cluster-api/controllers [sigs.k8s.io/cluster-api/controllers.test]
controllers/machinedeployment_controller_test.go:393:3: undefined: By
```

@fabriziopandini  @enxebre

**Comment by enxebre** [comment]
>Is it possibly related to this line in build logs? I did add a new test at line 393, but I don't quite understand why that would be the issue:
controllers/machinedeployment_controller_test.go:393:3: undefined: By

@Arvinderpal We are in the process of dropping ginkgo from the unit/integration tests. If you get the latest changes i.e `git fetch` this repo and `git rebase` your branch atop you'll see ginko is not anymore in that test package thus By is undefined.

File: controllers/machinedeployment_controller_test.go (hunk 5 above)
Comment (enxebre) [inline controllers/machinedeployment_controller_test.go:392]: we are not using ginkgo anymore here, please rebase atop main branch.

File: controllers/machinedeployment_sync_test.go [hunk 12]
```diff
@@ -397,3 +399,113 @@ func TestScaleMachineSet(t *testing.T) {
...
+
+func TestMachineDeploymentSyncDeploymentStatus(t *testing.T) {
+	pds := int32(60)
+	tests := []struct {
+		name               string
+		d                  *clusterv1.MachineDeployment
+		allMSs             []*clusterv1.MachineSet
+		newMS              *clusterv1.MachineSet
+		expectCond         bool
+		expectedCondType   clusterv1.ConditionType
```
Comment (enxebre) [inline controllers/machinedeployment_sync_test.go:461]: this does not seem to scale well as soon as we add another one condition. Actually is not enough to validate the two introduced by this PR.
Might want to change this to `conditions []*clusterv1.Condition`. The machine controller test has a small `assertConditions` library that might come in handy here.

File: controllers/machinedeployment_controller_test.go (hunk 4 above)
Comment (enxebre) [reply controllers/machinedeployment_controller_test.go:398]: Could you please point to other place where this is done so I can have a look? Regardless of where else it is used I see no value here.

File: controllers/machinedeployment_sync.go (hunk 2 above)
Comment (enxebre) [reply controllers/machinedeployment_sync.go:353]:
```
	// Rolling update config params. Present only if
	// MachineDeploymentStrategyType = RollingUpdate.
	// +optional
	RollingUpdate *MachineRollingUpdateDeployment `json:"rollingUpdate,omitempty"`
```

As today maxUnavailable and MaxSurge are only applicable when `// MachineDeploymentStrategyType = RollingUpdate.`

I think reinforcing this would make it easier for new folks to ramp up and avoid confusion.

File: api/v1alpha4/condition_consts.go [hunk 13]
```diff
@@ -201,3 +201,14 @@ const (
 	// from making any further remediations.
 	TooManyUnhealthyReason = "TooManyUnhealthy"
 )
+
+// Conditions and condition Reasons for  MachineDeployments
+
+const (
+	// MachineDeploymentAvailableCondition means the MachineDeployment is available, that is, at least the minimum available
+	// machines required (i.e. Spec.Replicas-MaxUnavailable) are up and running for at least minReadySeconds.
```
Comment (enxebre) [inline api/v1alpha4/condition_consts.go:209]: As per https://github.com/kubernetes-sigs/cluster-api/pull/4625/files#r635973403
`i.e. Spec.Replicas-MaxUnavailable when MachineDeploymentStrategyType = RollingUpdate.`

File: controllers/machinedeployment_controller_test.go (hunk 5 above)
Comment (Arvinderpal) [reply controllers/machinedeployment_controller_test.go]: Done. Thanks for pointing that out!

File: controllers/machinedeployment_sync_test.go (hunk 12 above)
Comment (Arvinderpal) [reply controllers/machinedeployment_sync_test.go]: Done.

File: api/v1alpha4/condition_consts.go (hunk 13 above)
Comment (Arvinderpal) [reply api/v1alpha4/condition_consts.go]: Done.

File: controllers/machinedeployment_sync.go (hunk 9 above)
Comment (fabriziopandini) [reply controllers/machinedeployment_sync.go:356]: Fine for me!

Comment (Arvinderpal) [reply controllers/machinedeployment_sync.go:356]: Cool. I updated the tracking issue with the new task.

**Comment by Arvinderpal** [comment]
@enxebre @fabriziopandini  I have addressed all your comments. Are we good to merge this? I will follow up with other conditions as documented in the tracker: https://github.com/kubernetes-sigs/cluster-api/issues/3486
CC @vincepri

File: controllers/machinedeployment_sync_test.go [hunk 14]
```diff
@@ -397,3 +398,109 @@ func TestScaleMachineSet(t *testing.T) {
...
+		},
+		Status: clusterv1.MachineSetStatus{
+			Selector:          "",
+			AvailableReplicas: availableReplicas,
+			Replicas:          statusReplicas,
+		},
+	}
+}
+
+func TestMachineDeploymentSyncDeploymentStatus(t *testing.T) {
```
Comment (enxebre) [inline controllers/machinedeployment_sync_test.go:452]: This should be named `TestSyncDeploymentStatus`

File: controllers/machinedeployment_sync_test.go [hunk 15]
```diff
@@ -397,3 +398,109 @@ func TestScaleMachineSet(t *testing.T) {
...
+				},
+			},
+		},
+		{
+			name:   "Deployment Available: MachineDeploymentAvailableCondition should exist and be true",
+			d:      currentDeployment(&pds, 3, 3, 3, 3, clusterv1.Conditions{}),
+			allMSs: []*clusterv1.MachineSet{},
+			newMS:  newMSWithAvailable("foo", 3, 3, 3),
+			expectedConditions: []*clusterv1.Condition{
+				&clusterv1.Condition{
```
Comment (enxebre) [inline controllers/machinedeployment_sync_test.go:481]: Shouldn't we expect a ready condition as well? https://github.com/kubernetes-sigs/cluster-api/pull/4625/files#diff-34f64847a48c9fc1b3222c9bec6c93f30767a93262215eb4305d78c3b9097c63R158

File: controllers/machinedeployment_sync_test.go [hunk 16]
```diff
@@ -397,3 +398,109 @@ func TestScaleMachineSet(t *testing.T) {
...
+		},
+	}
+}
+
+func TestMachineDeploymentSyncDeploymentStatus(t *testing.T) {
+	pds := int32(60)
+	tests := []struct {
+		name               string
+		d                  *clusterv1.MachineDeployment
+		allMSs             []*clusterv1.MachineSet
```
Comment (enxebre) [inline controllers/machinedeployment_sync_test.go:457]: Let's call this oldMachinesets, so allMachinesSets is what you get from `allMachineSets := append(tc.oldMachineSets, tc.newMachineSet)` consistently with other tests e.g `TestReconcileOldMachineSets`

File: controllers/machinedeployment_sync_test.go (hunk 15 above)
Comment (enxebre) [reply controllers/machinedeployment_sync_test.go:481]: mm just realised that happens at patching time.

**Comment by enxebre** [comment]
>@enxebre @fabriziopandini I have addressed all your comments. Are we good to merge this? I will follow up with other conditions as documented in the tracker: #3486
CC @vincepri

@Arvinderpal looks great to me overall, there's still some outstanding comments and tests and linters are failing.

File: controllers/machinedeployment_sync_test.go (hunk 14 above)
Comment (Arvinderpal) [reply controllers/machinedeployment_sync_test.go:452]: Done.

File: controllers/machinedeployment_sync_test.go (hunk 16 above)
Comment (Arvinderpal) [reply controllers/machinedeployment_sync_test.go]: Done

**Comment by Arvinderpal** [comment]
> @Arvinderpal looks great to me overall, there's still some outstanding comments and tests and linters are failing.

@enxebre Thanks. The linter error is not mine. It's an optional check so I'm guessing someone else merged with that failing. The apidiff-main is also optional and I believe is expected to fail given the api changes. 

```
api/v1alpha4/machinedeployment_webhook.go:47:29: ST1016: methods on the same type should have the same receiver name (seen 2x "in", 6x "m") (stylecheck)
func (m *MachineDeployment) Default() {
```

File: api/v1alpha4/condition_consts.go (hunk 3 above)
Comment (Arvinderpal) [reply api/v1alpha4/condition_consts.go:212]: Changed it to `NotMinimumMachinesAvailableReason` and updated the comment.

File: controllers/machinedeployment_sync.go (hunk 2 above)
Comment (Arvinderpal) [reply controllers/machinedeployment_sync.go:353]: Ok. Added the comment.

File: controllers/machinedeployment_sync_test.go [hunk 17]
```diff
@@ -397,3 +398,105 @@ func TestScaleMachineSet(t *testing.T) {
...
+		ObjectMeta: metav1.ObjectMeta{
+			Name:              name,
+			CreationTimestamp: metav1.Time{},
+			Namespace:         metav1.NamespaceDefault,
+		},
+		Spec: clusterv1.MachineSetSpec{
+			Replicas: pointer.Int32Ptr(specReplicas),
+		},
+		Status: clusterv1.MachineSetStatus{
+			Selector:          "",
```
Comment (enxebre) [inline controllers/machinedeployment_sync_test.go:445]: why do we need to set this selector?

File: controllers/machinedeployment_sync_test.go [hunk 18]
```diff
@@ -397,3 +398,105 @@ func TestScaleMachineSet(t *testing.T) {
 		})
 	}
 }
+
+func newDeploymentStatus(replicas, updatedReplicas, availableReplicas int32) clusterv1.MachineDeploymentStatus {
```
Comment (enxebre) [inline controllers/machinedeployment_sync_test.go:402]: what is this func giving us over just inlining in the one below?

File: controllers/machinedeployment_sync_test.go [hunk 19]
```diff
@@ -397,3 +398,105 @@ func TestScaleMachineSet(t *testing.T) {
...
+			},
+		},
+		Status: newDeploymentStatus(statusReplicas, updatedReplicas, availableReplicas),
+	}
+	d.Status.Conditions = conditions
+	return d
+}
+
+// helper to create MS with given availableReplicas.
+func newMSWithAvailable(name string, specReplicas, statusReplicas, availableReplicas int32) *clusterv1.MachineSet {
```
Comment (enxebre) [inline controllers/machinedeployment_sync_test.go:434]: `newTestMachinesetWithReplicas` may be is more meaningful?

File: controllers/machinedeployment_sync_test.go [hunk 20]
```diff
@@ -397,3 +398,105 @@ func TestScaleMachineSet(t *testing.T) {
...
+
+func newDeploymentStatus(replicas, updatedReplicas, availableReplicas int32) clusterv1.MachineDeploymentStatus {
+	return clusterv1.MachineDeploymentStatus{
+		Replicas:          replicas,
+		UpdatedReplicas:   updatedReplicas,
+		AvailableReplicas: availableReplicas,
+	}
+}
+
+func currentDeployment(pds *int32, replicas, statusReplicas, updatedReplicas, availableReplicas int32, conditions clusterv1.Conditions) *clusterv1.MachineDeployment {
```
Comment (enxebre) [inline controllers/machinedeployment_sync_test.go:410]: `newTestMachineDeployment` may be is more meaningful? the "current" temporary nuance make no sense within the function scope. It's up to the caller defining this into a variable to choose that nuance.

**Comment by enxebre** [comment]
>The linter error is not mine. It's an optional check so I'm guessing someone else merged with that failing. The apidiff-main is also optional and I believe is expected to fail given the api changes.
api/v1alpha4/machinedeployment_webhook.go:47:29: ST1016: methods on the same type should have the same receiver name (seen 2x "in", 6x "m") (stylecheck)
func (m *MachineDeployment) Default() {

mm I see ok hopefully someone more familiar with CI can validate this cc @sbueringer.

Dropped a few more small comments, other than that lgtm

File: api/v1alpha4/machinedeployment_types.go [hunk 21]
```diff
@@ -283,3 +287,11 @@ type MachineDeploymentList struct {
 func init() {
 	SchemeBuilder.Register(&MachineDeployment{}, &MachineDeploymentList{})
 }
+
+func (in *MachineDeployment) GetConditions() Conditions {
```
Comment (fabriziopandini) [inline api/v1alpha4/machinedeployment_types.go:291]: This is causing linter errors.
You should use m instead of in (same receiver name as existing methods)

File: api/v1alpha3/conversion.go [hunk 22]
```diff
@@ -158,6 +158,10 @@ func (dst *MachineDeployment) ConvertFrom(srcRaw conversion.Hub) error {
 	return nil
 }
 
+func Convert_v1alpha4_MachineDeploymentStatus_To_v1alpha3_MachineDeploymentStatus(in *v1alpha4.MachineDeploymentStatus, out *MachineDeploymentStatus, s apiconversion.Scope) error {
+	return autoConvert_v1alpha4_MachineDeploymentStatus_To_v1alpha3_MachineDeploymentStatus(in, out, s)
```
Comment (fabriziopandini) [inline api/v1alpha3/conversion.go:162]: We usually add a comment explaining why a custom conversion is required.
In this case it should be something like
// Status.Conditions was introduced in v1alpha4, thus requiring a custom conversion function; the values is going to be preserved in an annotation thus allowing roundtrip without loosing informations

File: api/v1alpha4/condition_consts.go [hunk 23]
```diff
@@ -201,3 +201,14 @@ const (
...
+
+// Conditions and condition Reasons for  MachineDeployments
+
+const (
+	// MachineDeploymentAvailableCondition means the MachineDeployment is available, that is, at least the minimum available
+	// machines required (i.e. Spec.Replicas-MaxUnavailable when MachineDeploymentStrategyType = RollingUpdate) are up and running for at least minReadySeconds.
+	MachineDeploymentAvailableCondition ConditionType = "Available"
+
+	// NotMinimumMachinesAvailableReason reflects the fact that the required minimum number of machines for a machinedeployment are not available.
+	NotMinimumMachinesAvailableReason = "NotMinimumMachinesAvailable"
```
Comment (fabriziopandini) [inline api/v1alpha4/condition_consts.go:213]: What about `WaitingForAvailableMachines`; also, should we drop available given that it is repetition considering it applies to `MachineDeploymentAvailableCondition`
Another alternative is to replace with something more user oriented, given that I'm not really sure the nuances of what available means are clear to the end users

**Comment by sbueringer** [comment]
@fabriziopandini  Apart from that you already found why the linter is failing, could it be that the golang-ci-lint check is currently not mandatory for merge?

/cc @enxebre

File: controllers/machinedeployment_sync_test.go (hunk 17 above)
Comment (Arvinderpal) [reply controllers/machinedeployment_sync_test.go:445]: not necessary. just copied pasted from other tests in this file. will remove it.

File: controllers/machinedeployment_sync_test.go (hunk 18 above)
Comment (Arvinderpal) [reply controllers/machinedeployment_sync_test.go:402]: removed it.

File: controllers/machinedeployment_sync_test.go (hunk 19 above)
Comment (Arvinderpal) [reply controllers/machinedeployment_sync_test.go:434]: done

File: controllers/machinedeployment_sync_test.go (hunk 20 above)
Comment (Arvinderpal) [reply controllers/machinedeployment_sync_test.go:410]: done

File: api/v1alpha4/machinedeployment_types.go (hunk 21 above)
Comment (Arvinderpal) [reply api/v1alpha4/machinedeployment_types.go:291]: That fixed. Thanks for pointing that out.

File: api/v1alpha3/conversion.go (hunk 22 above)
Comment (Arvinderpal) [reply api/v1alpha3/conversion.go:162]: done.

File: api/v1alpha4/condition_consts.go (hunk 23 above)
Comment (Arvinderpal) [reply api/v1alpha4/condition_consts.go:213]: seems like everyone has a different opinion on this :laughing: @fabriziopandini @enxebre I'm fine with what ever you guys agree on. Please let me know.

Comment (enxebre) [reply api/v1alpha4/condition_consts.go:213]: I'm good either way, let's go with @fabriziopandini suggestions if that make sense for everyone else
`WaitingForMachinesReason = WaitingForMachines`

Comment (Arvinderpal) [reply api/v1alpha4/condition_consts.go]: Done.

File: controllers/machinedeployment_sync_test.go [hunk 24]
```diff
@@ -397,3 +398,100 @@ func TestScaleMachineSet(t *testing.T) {
...
+				},
+			},
+		},
+		Status: clusterv1.MachineDeploymentStatus{
+			Replicas:          statusReplicas,
+			UpdatedReplicas:   updatedReplicas,
+			AvailableReplicas: availableReplicas,
+		},
+	}
+	d.Status.Conditions = conditions
```
Comment (srm09) [inline controllers/machinedeployment_sync_test.go:425]: qq: what does this line do? Was the idea to maybe test for modification on an existing set of conditions. If yes, seems like a good missing test case to cover.

File: api/v1alpha4/condition_consts.go [hunk 25]
```diff
@@ -201,3 +201,14 @@ const (
...
 )
+
+// Conditions and condition Reasons for  MachineDeployments
+
+const (
+	// MachineDeploymentAvailableCondition means the MachineDeployment is available, that is, at least the minimum available
+	// machines required (i.e. Spec.Replicas-MaxUnavailable when MachineDeploymentStrategyType = RollingUpdate) are up and running for at least minReadySeconds.
+	MachineDeploymentAvailableCondition ConditionType = "Available"
+
+	// WaitingForAvailableMachinesReason reflects the fact that the required minimum number of machines for a machinedeployment are not available.
```
Comment (srm09) [inline api/v1alpha4/condition_consts.go:212]: nit: Most other reason comments also add the **Severity** of the condition associated with the reason. Seems like a good idea to follow that practice.

File: controllers/machinedeployment_sync.go [hunk 26]
```diff
@@ -349,6 +350,16 @@ func (r *MachineDeploymentReconciler) scale(ctx context.Context, deployment *clu
...
+
+	// minReplicasNeeded will be equal to d.Spec.Replicas when the strategy is not RollingUpdateMachineDeploymentStrategyType.
+	minReplicasNeeded := *(d.Spec.Replicas) - mdutil.MaxUnavailable(*d)
+
+	if d.Status.AvailableReplicas >= minReplicasNeeded {
+		// NOTE: The structure of calculateStatus() does not allow us to update the machinedeployment directly, we can only update the status obj it returns. Ideally, we should change calculateStatus() --> updateStatus() to be consistent with the rest of the code base, until then, we update conditions here.
+		conditions.MarkTrue(d, clusterv1.MachineDeploymentAvailableCondition)
+	} else {
+		conditions.MarkFalse(d, clusterv1.MachineDeploymentAvailableCondition, clusterv1.WaitingForAvailableMachinesReason, clusterv1.ConditionSeverityWarning, "Minimum availability requires %d replicas, current %d available", minReplicasNeeded, d.Status.AvailableReplicas)
+	}
```
Comment (srm09) [inline controllers/machinedeployment_sync.go:362]: nit: Since the entire status is calculated in the `calculateStatus` method, and since Conditions are a part of the status, moving the condition generation logic to `calculateStatus` makes sense. That way, the entire status is re-calculated in a single method, and the name of the method encompasses the generation of the entire status.

File: controllers/machinedeployment_controller.go [hunk 27]
```diff
@@ -129,7 +130,12 @@ func (r *MachineDeploymentReconciler) Reconcile(ctx context.Context, req ctrl.Re
 
 	defer func() {
 		// Always attempt to patch the object and status after each reconciliation.
-		if err := patchHelper.Patch(ctx, deployment); err != nil {
+		// Patch ObservedGeneration only if the reconciliation completed successfully
+		patchOpts := []patch.Option{}
```
[the remark this answers is not in the document]
Comment (Arvinderpal) [reply controllers/machinedeployment_controller.go:134]: I see the := approach used far more frequently in the capi code base. If you don't mind, I will leave it as is.

File: controllers/machinedeployment_controller_test.go [hunk 28]
```diff
@@ -390,6 +390,18 @@ func TestMachineDeploymentReconciler(t *testing.T) {
...
+		t.Log("Verifying MachineDeployment has correct Conditions")
+		g.Eventually(func() bool {
+			key := client.ObjectKey{Name: deployment.Name, Namespace: deployment.Namespace}
+			g.Expect(testEnv.Get(ctx, key, deployment)).To(Succeed())
+			for _, c := range deployment.Status.Conditions {
+				if c.Type == clusterv1.MachineDeploymentAvailableCondition && c.Status == corev1.ConditionTrue {
+					return true
+				}
+			}
+			return false
```
[the remark this answers is not in the document]
Comment (Arvinderpal) [reply controllers/machinedeployment_controller_test.go:402]: Done.

File: controllers/machinedeployment_sync.go (hunk 26 above)
Comment (Arvinderpal) [reply controllers/machinedeployment_sync.go:362]: Please see the comment on line 358. This entire func will be refactored in a clean up PR.

File: controllers/machinedeployment_sync_test.go (hunk 24 above)
Comment (Arvinderpal) [reply controllers/machinedeployment_sync_test.go:425]: As we add more conditions to the machinedeployment, this will/may be useful. Please see tracking issue for additional conditions to be added: https://github.com/kubernetes-sigs/cluster-api/issues/3486

File: api/v1alpha4/condition_consts.go (hunk 25 above)
Comment (Arvinderpal) [reply api/v1alpha4/condition_consts.go:212]: Done.

**Review by Arvinderpal — COMMENTED** [review]
@srm09 Thanks for the feedback. I have addressed all the comments.

File: controllers/machinedeployment_sync_test.go (hunk 24 above)
Comment (enxebre) [reply controllers/machinedeployment_sync_test.go:425]: Please inline this within the the rest of the status unless there's a reason to not, otherwise it suggests there's some nuance the reader is missing and results confusing.

```
		Status: clusterv1.MachineDeploymentStatus{
			Replicas:          statusReplicas,
			UpdatedReplicas:   updatedReplicas,
			AvailableReplicas: availableReplicas,
                         Conditions: conditions,
		},
```

>As we add more conditions to the machinedeployment, this will/may be useful.

Also as general rule please let's don't over abstract ahead of time as it makes things unnecessarily more difficult to reason about when looking at PRs, code snapshots, git history... Let's abstract when we need it if we need it.

**Comment by enxebre** [comment]
other than https://github.com/kubernetes-sigs/cluster-api/pull/4625/files#r640464110
/lgtm
/hold
For someone to validate why tests are failing.
Thanks for patiently addressing all feedback @Arvinderpal. As discussed in the community call we'll wait post v1alpha4 to merge so we have time to get to the final of the effort and ship all conditions in one release.

File: controllers/machinedeployment_sync_test.go (hunk 24 above)
Comment (Arvinderpal) [reply controllers/machinedeployment_sync_test.go]: done.

**Comment by Arvinderpal** [comment]
@enxebre I have addressed the last comment. 
Waiting for v1alpha4 sounds good to me.

**Comment by Arvinderpal** [comment]
Now that v1alpha4 is out, can we merge this?

**Comment by Arvinderpal** [comment]
@fabriziopandini @srm09 This has been sitting for a while. If we can merge this, then I can follow up with [additional Conditions](https://github.com/kubernetes-sigs/cluster-api/issues/3486). :pray:

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
/approve

**Comment by CecileRobertMichon** [comment]
`FAIL	sigs.k8s.io/cluster-api/controllers [build failed]`

?

**Comment by Arvinderpal** [comment]
> `FAIL sigs.k8s.io/cluster-api/controllers [build failed]`

@CecileRobertMichon Can't seem to make out from the logs why this is happening. It could just be flaky so I'll try retest again but if that fails, I can try to rebase my branch. This will remove all the lgtms and approves...

**Comment by Arvinderpal** [comment]
@CecileRobertMichon @enxebre @srm09 Please lgtm and approve again. I had to rebase with master in order for all CI jobs to pass. :pray:
